bpo-31334: Fix timeout in select.poll.poll()

volans- opened the pull request:
Always pass `-1`, or `INFTIM` where defined, to the `poll()` system call when
a negative timeout is passed to the `poll.poll([timeout])` method in the
`select` module. Various OSes throw an error with arbitrary negative
values.

<!-- issue-number: bpo-31334 -->
https://bugs.python.org/issue31334
<!-- /issue-number -->

volans- committed:
bpo-31334: Fix timeout in select.poll.poll()

Always pass -1, or INFTIM where defined, to the poll() system call when
a negative timeout is passed to the poll.poll([timeout]) method in the
select module. Various OSes throw an error with arbitrary negative
values.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

volans- commented:
For the record I've signed the CLA and my status has been updated on bugs.python.org.

berkerpeksag commented on Modules/selectmodule.c:
```diff
@@ -554,7 +555,14 @@ poll_poll(pollObject *self, PyObject *args)
+        if (ms >= 0)
+            deadline = _PyTime_GetMonotonicClock() + timeout;
+        else    /* On many OSes timeout must be -1 or INFTIM, issue 31334 */
+#ifdef INFTIM
```
Can't we do this in the following block?

```c
if (timeout_obj == NULL || timeout_obj == Py_None) {
    timeout = -1;
#ifdef INFTIM
    ms = INFTIM;
#else
    ms = -1;
#endif
    deadline = 0;
}
```

Then we can just change

```c
deadline = _PyTime_GetMonotonicClock() + timeout;
```

to

```c
if (ms >= 0) {
    deadline = _PyTime_GetMonotonicClock() + timeout;
}
```

volans- replied:
I don't see how what you're proposing can fix the issue. When a user passes a negative `timeout` parameter it enters the `else` block at line 542, so `ms` will be negative and the system call will still fail. The problem is not the `deadline` but the `ms` variable.

But your comment made me realise that I should add the `#ifdef` to the `if` block too, for coherence and in case an OS will ever have `INFTIM != -1`. To be DRY I'll move the `#ifdef` after the `else` block.

berkerpeksag commented on Modules/selectmodule.c:
```diff
@@ -532,11 +532,12 @@ poll_poll(pollObject *self, PyObject *args)
     if (timeout_obj == NULL || timeout_obj == Py_None) {
         timeout = -1;
         ms = -1;
-        deadline = 0;   /* initialize to prevent gcc warning */
```
See my comment about `INFTIM` below. I think we don't need to move this line out of this `if` branch.

berkerpeksag commented on Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst:
```diff
@@ -0,0 +1,2 @@
+Fix ``poll.poll([timeout])`` in the ``select`` module for arbitrary negative
+timeouts on all OSes where it can only be a non-negative integer or -1.
```
Please add "Patch by Riccardo Coccioli.".

volans- replied:
Ok, thanks.

berkerpeksag commented on Lib/test/test_poll.py:
```diff
@@ -130,7 +130,7 @@ def test_poll2(self):
         pollster = select.poll()
         pollster.register( p, select.POLLIN )
-        for tout in (0, 1000, 2000, 4000, 8000, 16000) + (-1,)*10:
+        for tout in (0, 1000, 2000, 4000, 8000, 16000, -1000) + (-1,)*10:
```
Style nitpick: I'd move `-1000` to the beginning of the tuple to make it ordered.

volans- replied:
Ok.

berkerpeksag commented on Modules/selectmodule.c:
```diff
@@ -532,11 +532,12 @@ poll_poll(pollObject *self, PyObject *args)
-        deadline = 0;   /* initialize to prevent gcc warning */
     }
     else {
         if (_PyTime_FromMillisecondsObject(&timeout, timeout_obj,
```
Another question: Can't we just set `timeout` to `INFTIM` or `-1` if `timeout_obj` is less than `-1` instead of messing with `ms`?

volans- replied:
First of all thanks for the review.

I'm not getting why this would be better. The current checks against `INT_MIN` and `INT_MAX` are done with `ms`, so I followed that pattern. Also `ms` is the actual parameter passed to the system call and the one causing the issue right now, it seemed more logic to me to change directly this one.

volans- committed:
bpo-31334: Address some PR comments and simplify

volans- commented:
@berkerpeksag did you had any chance to look at my replies and changes?

volans- commented:
I didn't expect the Spanish Inquisition!

volans- commented:
@berkerpeksag do you have any additional feedback on this change?

berkerpeksag commented:
@volans- sorry for the delay. I need to take a closer look at this (and a BSD to test it locally) Please ping me again if I don't get to it by the next week.

berkerpeksag commented on Modules/selectmodule.c:
```diff
@@ -532,11 +532,12 @@ poll_poll(pollObject *self, PyObject *args)
         return NULL;
     }
 
+    deadline = 0;   /* initialize to prevent gcc warning */
```
We can set this in `deadline` initialization:

```c
_PyTime_t deadline = 0;
```

berkerpeksag replied:
And the following comment is not needed anymore:

```c
/* initialize to prevent gcc warning */
```

berkerpeksag commented on Modules/selectmodule.c:
```diff
@@ -532,11 +532,12 @@ poll_poll(pollObject *self, PyObject *args)
+    deadline = 0;   /* initialize to prevent gcc warning */
+
     /* Check values for timeout */
     if (timeout_obj == NULL || timeout_obj == Py_None) {
```
Looking at this again, we can get rid of this if branch if we move out default values to variable declarations:

```c
_PyTime_t timeout = -1, ms = -1, deadline = 0;
```

Then we can get rid of the following comment too:

```c
/* Check values for timeout */
```

berkerpeksag commented on Modules/selectmodule.c:
```diff
@@ -554,9 +555,17 @@ poll_poll(pollObject *self, PyObject *args)
         }
 
-        deadline = _PyTime_GetMonotonicClock() + timeout;
+        if (timeout >= 0)
```
Style nit: Per PEP 7, we now use the following style in new C code:

```c
if (spam) {
    return 42;
}
````

instead of

```c
if (spam)
    return 42;
````

berkerpeksag commented on Modules/selectmodule.c:
```diff
@@ -554,9 +555,17 @@ poll_poll(pollObject *self, PyObject *args)
+            deadline = _PyTime_GetMonotonicClock() + timeout;
     }
 
+    if (ms < 0) /* On many OSes timeout must be INFTIM or -1, issue 31334 */
```
Style nit: Same as above.

berkerpeksag replied:
Can we make this comment clearer? For example, it would be nice to mention this is the case if negative timeout is less than -1 on some OSes.

Also, I'd replace "many" with "some" or "some BSD OSes" since this is the first time this issue has been reported.

berkerpeksag commented on Lib/test/test_poll.py:
```diff
@@ -130,7 +130,7 @@ def test_poll2(self):
         pollster = select.poll()
         pollster.register( p, select.POLLIN )
-        for tout in (0, 1000, 2000, 4000, 8000, 16000) + (-1,)*10:
+        for tout in (-1000, 0, 1000, 2000, 4000, 8000, 16000) + (-1,)*10:
```
This is not directly related to this ticket, but I realized that `None` is not tested in `test_poll`. Could you also add `None` while we are here?

volans- committed:
bpo-31334: Address PR comments

volans- commented:
I have made the requested changes; please review again.

@berkerpeksag : Thank you very much for the review and the suggestions to simplify the code. My first patch was as less impactful as possible because I've found in other communities harder to get patches approved if they were also refactoring existing code. I'm happy to see a different approach here.

berkerpeksag commented on Modules/selectmodule.c:
```diff
@@ -554,7 +548,20 @@ poll_poll(pollObject *self, PyObject *args)
+    /* On some OSes, typically BSD-based ones, the timeout parameter of the
+       poll() syscall, when negative, must be exactly INFTIM, where defined,
+       or -1. See issue 31334. */
+    if (ms < 0) {
```
Nitpick: Can we change this to `ms < -1`? Since we set `ms = -1` in line 528, we shouldn't get any "invalid argument" exception if we don't pass a `timeout` value to `p.poll()`.

volans- replied:
@berkerpeksag what if a OS defines `INFTIM` different from `-1`? This ensures that we set `ms` to `INFTIM` in that case too. I can change it to `ms <= -1` if you think it's more readable.

On a side note, the documentation says only that the timeout parameter is in milliseconds, but doesn't say it must be an integer. I've noticed that passing a float works too and in the special case of passing a float in the range `(-1, 0)`, that is converted to `0` by `ms = _PyTime_AsMilliseconds(timeout, _PyTime_ROUND_CEILING);`. I've tested that this behaviour does not change with my patch, but is arguably the desired one.

serhiy-storchaka replied:
I concur with @volans-. It is better to keep `ms < 0`.

As for rounding problem, I just have opened bpo-31786. I didn't mentioned it here because seems this change doesn't change this behavior.

berkerpeksag replied:
@volans- fair enough, thanks for the explanation. I think `ms < 0` can stay as is, no need to change it to `ms <= -1`.

berkerpeksag reviewed (approved):
This looks pretty good to me, thank you! I just left a comment about the `ms < 0` check.

I will ask for another pair of eyes from another core developer then merge it.

serhiy-storchaka commented on Lib/test/test_poll.py:
```diff
@@ -130,7 +130,7 @@ def test_poll2(self):
         pollster = select.poll()
         pollster.register( p, select.POLLIN )
-        for tout in (0, 1000, 2000, 4000, 8000, 16000) + (-1,)*10:
+        for tout in (None, -1000, 0, 1000, 2000, 4000, 8000, 16000) + (-1,)*10:
```
I'm not sure that it is correct to insert infinite timeouts before finite timeouts in this test. Add them after `16000`. And maybe correct the multiplier in `(-1,)*10`.

volans- committed:
bpo-31334: Address PR comments

* Move infinite timeouts at the end of the tuple in the test.
* Add a timeout smaller than the command output generation between the
  first one and the second one (500).
* Adjusted the multiplier of the timeout (-1,) to 3, so to have only
  one additional call to poll after the last output from the command.

volans- commented:
I have made the requested changes; please review again.

@serhiy-storchaka Thanks for the review. I hope that I correctly understood what you meant with:
> correct the multiplier in `(-1,)*10`.

I've also added a timeout smaller than the command output generation between the first one and the second one (`500`) and se the multiplier for the `-1` to `3`.
The result is that in one case the timeout ends before any output is generated (`500`), and in one case (the last `-1`) it checks that there is no output once the pipe is closed. In all the other cases one line of output is read.

serhiy-storchaka commented on Lib/test/test_poll.py:
```diff
@@ -130,7 +130,7 @@ def test_poll2(self):
         pollster = select.poll()
         pollster.register( p, select.POLLIN )
-        for tout in (0, 1000, 2000, 4000, 8000, 16000) + (-1,)*10:
+        for tout in (0, 500, 1000, 2000, 4000, 8000, 16000, None, -1000) + (-1,)*3:
```
I think that the total number of infinity timeouts should be not less than 10.  There are no guaranties that all finite timeouts are not expired before the subprocess will write the first line.

volans- replied:
@serhiy-storchaka Oh, sorry, now I got what you meant, updating the patch.

volans- committed:
bpo-31334: Fix test

berkerpeksag commented:
@serhiy-storchaka can you take a look at the last changes? I'd like to merge this if everything looks good to you.

serhiy-storchaka commented:
An explicit test for infinity timeout will be added in #4003. It looks a better place for testing this PR. I suggest to wait until #4003 is merged.

vstinner commented on Modules/selectmodule.c:
```diff
@@ -554,7 +548,20 @@ poll_poll(pollObject *self, PyObject *args)
+        ms = INFTIM;
+#else
+        ms = -1;
+#endif
```
This normalization must be moved at the beginning of the loop below, since ms is recomputed if poll() is interrupted by a signal (EINTR).

volans- replied:
@haypo where do you mean exactly? `ms` is recalculated only if `timeout` is non-negative and `ms` cannot be negative with a non-negative `timeout`.

serhiy-storchaka replied:
I agree with @volans-. No changes are needed here.

vstinner replied:
Right.

vstinner commented on Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst:
```diff
@@ -0,0 +1,3 @@
+Fix ``poll.poll([timeout])`` in the ``select`` module for arbitrary negative
+timeouts on all OSes where it can only be a non-negative integer or -1.
```
nitpick: "non-negative" is better known as "positive" :-)

volans- replied:
@haypo  "positive" doesn't include zero... :-)

vstinner replied:
"strictly positive" doesn't include zero, "positive" does. Well, I don't really care. The current sentence is correct.

serhiy-storchaka replied:
"non-negative" means "positive or zero".

vstinner reviewed (approved):
LGTM, except of a minor nitpick.

serhiy-storchaka commented on Lib/test/test_poll.py:
```diff
@@ -130,7 +130,7 @@ def test_poll2(self):
         pollster = select.poll()
         pollster.register( p, select.POLLIN )
-        for tout in (0, 1000, 2000, 4000, 8000, 16000) + (-1,)*10:
+        for tout in (0, 500, 1000, 2000, 4000, 8000, 16000, None, -1000) + (-1,)*8:
```
#4003 is merged. Please revert this change and add the timeout value -1000 in the new test method.

volans- replied:
@serhiy-storchaka sure, will do.
What is the preferred method to update this branch with the new upstream code, `git merge`? I'm asking because this is my first patch and the developer guide mention to not use _anything that would require a force push to GitHub_, so I guess a rebase is excluded.

serhiy-storchaka replied:
Yes, just use `git merge`.

volans- committed:
Merge remote-tracking branch 'upstream/master' into fix-select-poll

volans- committed:
bpo-31334: use new test function

* revert changes to test_poll2 to use the new dedicated function
  test_poll_blocks_with_negative_ms to test negative values.

miss-islington commented:
Thanks @volans- for the PR, and @haypo for merging it 🌮🎉.. I'm working now to backport this PR to: 2.7, 3.6.
🐍🍒⛏🤖

vstinner commented:
Wow, 3 approvals from 3 different core developers. Well done @volans- ! Thanks for this nice bug fix ;-)

miss-islington commented:
Sorry, @volans- and @haypo, I could not cleanly backport this to `3.6` due to a conflict. 
                            Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line.
                            ```
                            cherry_picker 6cfa927ceb931ad968b5b03e4a2bffb64a8a0604 3.6
                            ```

miss-islington commented:
Sorry, @volans- and @haypo, I could not cleanly backport this to `2.7` due to a conflict. 
                            Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line.
                            ```
                            cherry_picker 6cfa927ceb931ad968b5b03e4a2bffb64a8a0604 2.7
                            ```

vstinner commented:
I removed the "needs backport" labels to prevent the bot to try again to create backport fixes.

volans- commented:
@haypo thank you all for the review, comments and for approving/merging it.

Should I create manual PRs for the backports given that the bot seems to have some issues today?

serhiy-storchaka commented:
First #4003 should be backported.

miss-islington commented:
Thanks @volans- for the PR, and @haypo for merging it 🌮🎉.. I'm working now to backport this PR to: 3.6.
🐍🍒⛏🤖

miss-islington commented:
Sorry, @volans- and @haypo, I could not cleanly backport this to `3.6` due to a conflict. 
                            Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line.
                            ```
                            cherry_picker 6cfa927ceb931ad968b5b03e4a2bffb64a8a0604 3.6
                            ```

miss-islington commented:
Thanks @volans- for the PR, and @haypo for merging it 🌮🎉.. I'm working now to backport this PR to: 3.6.
🐍🍒⛏🤖

miss-islington commented:
Thanks @volans- for the PR, and @haypo for merging it 🌮🎉.. I'm working now to backport this PR to: 2.7.
🐍🍒⛏🤖 I'm not a witch! I'm not a witch!

miss-islington commented:
Sorry, @volans- and @haypo, I could not cleanly backport this to `2.7` due to a conflict. 
                            Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line.
                            ```
                            cherry_picker 6cfa927ceb931ad968b5b03e4a2bffb64a8a0604 2.7
                            ```

volans- commented:
@serhiy-storchaka I've ported this to the 2.7 branch with GH-4034, it's a smaller patch given the code in the 2.7 branch.

serhiy-storchaka commented:
Thanks @volans-!
